xarray/core/dataset.py

```diff
@@ -6835,10 +6836,13 @@ def reduce(
                 if (
                     # Some reduction functions (e.g. std, var) need to run on variables
                     # that don't have the reduce dims: PR5393
-                    not reduce_dims
-                    or not numeric_only
-                    or np.issubdtype(var.dtype, np.number)
-                    or (var.dtype == np.bool_)
+                    (
+                        not reduce_dims
+                        or not numeric_only
+                        or np.issubdtype(var.dtype, np.number)
+                        or (var.dtype == np.bool_)
+                    )
+                    and not is_extension_array_dtype(var.dtype)
```

Comment:
What happens if there's an ExtensionArray data var in a dataset and you call `ds.mean()`? Are we silently dropping or raising an error?

I think we should raise a nice error and ask the user to drop it themselves.

Reply:
Ok, but this is different from the behavior with a numpy object array, isn't it?

```python
import xarray as xr
import numpy as np
import pandas as pd

ds = xr.Dataset({'cat': (['x'], np.array(['a', 'b', 'c'])), 'num': (['x'], np.array([1, 2, 3]))}, coords={'x': ('x', [0, 1, 2])})
ds.mean() # drops `cat`
```

Reply:
Good point. Thanks!

Reply:
I guess we'd want to be careful about allowing Extension Arrays with ints, floats, datetimes etc. looking at the list here: https://pandas.pydata.org/docs/reference/api/pandas.array.html but this can be a followup.

Reply:
Agreed.  For now I will implement the drop and not error then.

Reply:
I think this is fine for now?

Reply:
@dcherian implemented, but is there a test case for dropping non-numerics I can add this to?  I couldn't find one

Reply:
https://github.com/pydata/xarray/blob/cf3655968b8b12cc0ecd28fb324e63fb94d5e7e2/xarray/tests/test_dataset.py#L5459